### ISSUE TITLE
Finish event not triggered

### ISSUE THREAD

**Issue by Killian-denechere** [katspaugh/wavesurfer.ts#86]
Hi !

I've bound a `finish` event to my wavesurfer instance like this :
```js
const wavesurfer = WaveSurfer.create({ ... });
wavesurfer.on("finish", () => console.log('The audio playing is over'));
```
but, when my audio comes to the end, while the cursor is stopped, the log message is not displayed in my console.

I think the event is not thrown at all.
I've searched though the examples and repo issues, but didn't find anything about this event.
I tried it out on the Wavesurfer.js version and it's working as expected here.

Also, as the event `error` is available in Wavesurfer.js, do you plan to add this event on this Typescript version ?

Thank you for your answers 🙂

**Comment by Kevin-Android** [comment]
> 

I have it working fine in my code example, the finish event callback works fine, what version are you using?

**Comment by katspaugh** [comment]
For me it also works fine on the audio files I tested. I've just checked it in Chrome, Safari and Firefox.
Added an example for this event here: https://wavesurfer-ts.pages.dev/#timeline.js

We can't exclude that it might not work with some audio formats/specific files, so please share your code/audio file.

**Comment by katspaugh** [comment]
Tried with this file, without and without peaks, it emits `finish` in both cases.

<img width="1535" alt="Screenshot 2023-05-24 at 20 22 18" src="https://github.com/katspaugh/wavesurfer.ts/assets/381895/d3aadb5c-eca8-4709-961e-5d49cc3f747d">

Can you reproduce it with this code?
```
// With pre-decoded audio data

import WaveSurfer from 'https://unpkg.com/wavesurfer.js@alpha'

const wavesurfer = WaveSurfer.create({
  container: document.body,
  waveColor: 'rgb(200, 0, 200)',
  progressColor: 'rgb(100, 0, 100)',
  normalize: true,
})

fetch('/examples/audio/record.json').then((response) => response.json()).then((json) => {
  wavesurfer.load(
    '/examples/audio/record.mp3',
    json.data
  )
})

wavesurfer.on('interaction', () => {
  wavesurfer.play()
})

wavesurfer.on('finish', () => {
  console.log('Finished!')
})
```

**Comment by katspaugh** [comment]
Thanks for checking!
Yeah, HTTP servers can be tricky when it comes to media file serving, so I'm glad you've nailed that down.

> When playing a mp3 audio file without peaks, no problem, the event is thrown
> When playing a mp3 audio file with peaks, I can seek in the audio but as soon as the playing cursor comes to the end, the audio is stopped, the cursor also, but the event is not thrown

With pre-decoded peaks, one big difference is that it will always use the audio's duration as reported by the HTML audio element. It's known to be problematic for some files. E.g. an inaccurate value is returned or even Infinity.
Without peaks, it will fall back to the duration detected by the Web Audio decoder.

So I would suggest to try the following:

```
const wavesurfer = WaveSurfer.create({ ... })

// Access the internally created audio element
const audio = wavesurfer.getMediaElement()

// Check what duration it has
audio.addEventListener(''loadedmetadata', () => {
  console.log('Duration', audio.duration)
})
```

If it gives you something weird for the duration, the finish event won't know when to fire.

**Comment by katspaugh** [comment]
That said, I've just realized there's a more robust way to emit the finish event, namely the [`ended` event](https://html.spec.whatwg.org/multipage/media.html#event-media-ended):
<img width="1125" alt="Screenshot 2023-05-25 at 07 23 19" src="https://github.com/katspaugh/wavesurfer.ts/assets/381895/cff36a49-11f0-4977-bebd-2800e76da329">

I'm going to make a PR to use that.

**Comment by katspaugh** [comment]
Makes sense, thanks for verifying. With the fix in #89 it should work without passing a duration. It now relies on the native ended event.
